apps/src/lib/node/ledger/shell/mod.rs

```diff
@@ -849,32 +849,71 @@ where
                      blocks"
                 ),
             }
+
+            // broadcast any queued txs
+            self.broadcast_queued_txs();
         }
 
-        #[cfg(not(feature = "abcipp"))]
-        {
-            use crate::node::ledger::shell::vote_extensions::iter_protocol_txs;
+        response
+    }
 
-            if let ShellMode::Validator { .. } = &self.mode {
-                let ext = self.craft_extension();
+    /// Empties all the ledger's queues of transactions to be broadcasted
+    /// via CometBFT's P2P network.
+    #[inline]
+    fn broadcast_queued_txs(&mut self) {
+        self.broadcast_protocol_txs();
+        self.broadcast_expired_txs();
+    }
 
-                let protocol_key = self
-                    .mode
-                    .get_protocol_key()
-                    .expect("Validators should have protocol keys");
+    /// Broadcast any pending protocol transactions.
+    fn broadcast_protocol_txs(&mut self) {
+        use crate::node::ledger::shell::vote_extensions::iter_protocol_txs;
 
-                let protocol_txs = iter_protocol_txs(ext).map(|protocol_tx| {
-                    protocol_tx
-                        .sign(protocol_key, self.chain_id.clone())
-                        .to_bytes()
-                });
+        let ext = self.craft_extension();
 
-                for tx in protocol_txs {
-                    self.mode.broadcast(tx);
-                }
-            }
+        let protocol_key = self
+            .mode
+            .get_protocol_key()
+            .expect("Validators should have protocol keys");
+
+        let protocol_txs = iter_protocol_txs(ext).map(|protocol_tx| {
+            protocol_tx
+                .sign(protocol_key, self.chain_id.clone())
+                .to_bytes()
+        });
+
+        for tx in protocol_txs {
+            self.mode.broadcast(tx);
+        }
+    }
+
+    /// Broadcast any expired transactions.
+    fn broadcast_expired_txs(&mut self) {
+        let eth_events = {
+            let mut events: Vec<_> = self
+                .wl_storage
+                .storage
+                .expired_txs_queue
+                .drain()
+                .map(|expired_tx| match expired_tx {
+                    ExpiredTx::EthereumEvent(event) => event,
+                })
+                .collect();
+            events.sort();
```

Comment:
what's the point of sorting them?

Reply:
I think this was about being absolutely deterministic in the payload you signed. can't remember the exact details of this tbh. maybe it made more sense when we had digests of events

Reply:
so the digests were used to reconstruct the vote extensions from every validator, signature pair stored within. I think this was the reason we required events to be sorted, so we had signatures over the same data. it's probably not required with individual vote extensions, for them to be signed